File: summer/WebContent/js/MS/internal/data/CommitManager.js
/**
 * CommitManager
 */

define(["dojo/_base/declare", "system/Type", "generic/Dictionary", "collections/IEnumerable",
        "generic/List"], function(declare, Type, Dictionary, IEnumerable){
//	class 
	var Set = declare([Dictionary, IEnumerable], {
   	 	"-chains-": {
   	 		constructor: "manual"
    	},
        constructor:function(/*IDictionary*/ other) 
        {
        	if(other != undefined){
        		Dictionary.prototype.constructor.call(this.other);
        	}
        },

//        public void 
		Add:function(/*T*/ item) 
        { 
            if (!this.ContainsKey(item))
            { 
            	this.Add(item, null);
            }
        },

//        IEnumerator<T> IEnumerable<T>.
        GetEnumerator:function()
        { 
            return this.Keys.GetEnumerator(); 
        },

//        public List<T> 
        ToList:function()
        {
            return new List(this.Keys);
        } 
    });
	
//    static readonly List<BindingGroup>
	var EmptyBindingGroupList = new List/*<BindingGroup>*/(); 
//    static readonly List<BindingExpressionBase> 
	var EmptyBindingList = new List/*<BindingExpressionBase>*/();
    
	var CommitManager = declare("CommitManager", null,{
		constructor:function( ){
//	        Set<BindingGroup> 
			this._bindingGroups = new Set/*<BindingGroup>*/(); 
//	        Set<BindingExpressionBase> 
			this._bindings = new Set/*<BindingExpressionBase>*/(); 
		},
		

//        internal void 
        AddBindingGroup:function(/*BindingGroup*/ bindingGroup) 
        { 
            this._bindingGroups.Add(bindingGroup);
        },

//        internal void 
        RemoveBindingGroup:function(/*BindingGroup*/ bindingGroup)
        {
        	this._bindingGroups.Remove(bindingGroup); 
        },
 
//        internal void 
        AddBinding:function(/*BindingExpressionBase*/ binding) 
        {
        	this._bindings.Add(binding); 
        },

//        internal void 
        RemoveBinding:function(/*BindingExpressionBase*/ binding)
        { 
        	this._bindings.Remove(binding);
        }, 
 
//        internal List<BindingGroup> 
        GetBindingGroupsInScope:function(/*DependencyObject*/ element)
        { 
            // iterate over a copy of the full list - callouts can
            // change the original list
            /*List<BindingGroup>*/var fullList = this._bindingGroups.ToList();
            /*List<BindingGroup>*/var list = EmptyBindingGroupList; 

            for (var i=0; i< fullList.Count; i++) 
            { 
            	/*BindingGroup*/var bindingGroup = fullList.Get(i);
                /*DependencyObject*/var owner = bindingGroup.Owner;
                if (owner != null && IsInScope(element, owner)) 
                {
                    if (list == EmptyBindingGroupList)
                    {
                        list = new List/*<BindingGroup>*/(); 
                    }
 
                    list.Add(bindingGroup); 
                }
            } 

            return list;
        },
 
//        internal List<BindingExpressionBase> 
        GetBindingsInScope:function(/*DependencyObject*/ element)
        { 
            // iterate over a copy of the full list - calling TargetElement can 
            // change the original list
            /*List<BindingExpressionBase>*/var fullList = this._bindings.ToList(); 
            /*List<BindingExpressionBase>*/var list = EmptyBindingList;

            for ( var i=0; i<fullList.Count; i++)
            { 
            	/*BindingExpressionBase*/var binding = fullList.Get(i);
                /*DependencyObject*/var owner = binding.TargetElement;
                if (owner != null && 
                    binding.IsEligibleForCommit && 
                    this.IsInScope(element, owner))
                { 
                    if (list == EmptyBindingList)
                    {
                        list = new List/*<BindingExpressionBase>*/();
                    } 

                    list.Add(binding); 
                } 
            }
 
            return list;
        },

        // remove stale entries 
//        internal bool 
        Purge:function()
        { 
//            bool foundDirt = false; 
//
//            int count = _bindings.Count; 
//            if (count > 0)
//            {
//                List<BindingExpressionBase> list = _bindings.ToList();
//                foreach (BindingExpressionBase binding in list) 
//                {
//                    // fetching TargetElement may detach the binding, removing it from _bindings 
//                    DependencyObject owner = binding.TargetElement; 
//                }
//            } 
//            foundDirt = foundDirt || (_bindings.Count < count);
//
//            count = _bindingGroups.Count;
//            if (count > 0) 
//            {
//                List<BindingGroup> list = _bindingGroups.ToList(); 
//                foreach (BindingGroup bindingGroup in list) 
//                {
//                    // fetching Owner may detach the binding group, removing it from _bindingGroups 
//                    DependencyObject owner = bindingGroup.Owner;
//                }
//            }
//            foundDirt = foundDirt || (_bindingGroups.Count < count); 
//
//            return foundDirt; 
        },

        // return true if element is a descendant of ancestor 
//        bool 
        IsInScope:function(/*DependencyObject*/ ancestor, /*DependencyObject*/ element)
        { 
            /*bool*/var result = (ancestor == null) || VisualTreeHelper.IsAncestorOf(ancestor, element); 
            return result;
        } 
	});
	
	Object.defineProperties(CommitManager.prototype,{
//		internal bool 
		IsEmpty:
        {
            get:function() {
            	return this._bindings.Count == 0 && this._bindingGroups.Count == 0; 
            }
        } 
	});
	
	CommitManager.Type = new Type("CommitManager", CommitManager, [Object.Type]);
	return CommitManager;
});





        

File: org.summer.view.widget/src/org/summer/view/widget/internal/CommitManager.java
package org.summer.view.widget.internal;

import java.util.List;

import org.summer.view.widget.DependencyObject;
import org.summer.view.widget.data.BindingExpressionBase;
import org.summer.view.widget.data.BindingGroup;

/*internal*/public class CommitManager 
{ 
//    #region Internal Methods

    /*internal*/ boolean IsEmpty
    {
        get { return _bindings.Count == 0 && _bindingGroups.Count == 0; }
    } 

    /*internal*/ public void AddBindingGroup(BindingGroup bindingGroup) 
    { 
        _bindingGroups.Add(bindingGroup);
    } 

    /*internal*/ public void RemoveBindingGroup(BindingGroup bindingGroup)
    {
        _bindingGroups.Remove(bindingGroup); 
    }

    /*internal*/ public void AddBinding(BindingExpressionBase binding) 
    {
        _bindings.Add(binding); 
    }

    /*internal*/ public void RemoveBinding(BindingExpressionBase binding)
    { 
        _bindings.Remove(binding);
    } 

    /*internal*/ public List<BindingGroup> GetBindingGroupsInScope(DependencyObject element)
    { 
        // iterate over a copy of the full list - callouts can
        // change the original list
        List<BindingGroup> fullList = _bindingGroups.ToList();
        List<BindingGroup> list = EmptyBindingGroupList; 

        for/*each*/ (BindingGroup bindingGroup : fullList) 
        { 
            DependencyObject owner = bindingGroup.Owner;
            if (owner != null && IsInScope(element, owner)) 
            {
                if (list == EmptyBindingGroupList)
                {
                    list = new List<BindingGroup>(); 
                }

                list.Add(bindingGroup); 
            }
        } 

        return list;
    }

    /*internal*/ public List<BindingExpressionBase> GetBindingsInScope(DependencyObject element)
    { 
        // iterate over a copy of the full list - calling TargetElement can 
        // change the original list
        List<BindingExpressionBase> fullList = _bindings.ToList(); 
        List<BindingExpressionBase> list = EmptyBindingList;

        for/*each*/ (BindingExpressionBase binding : fullList)
        { 
            DependencyObject owner = binding.TargetElement;
            if (owner != null && 
                binding.IsEligibleForCommit && 
                IsInScope(element, owner))
            { 
                if (list == EmptyBindingList)
                {
                    list = new List<BindingExpressionBase>();
                } 

                list.Add(binding); 
            } 
        }

        return list;
    }

    // remove stale entries 
    /*internal*/ public boolean Purge()
    { 
        boolean foundDirt = false; 

        int count = _bindings.Count; 
        if (count > 0)
        {
            List<BindingExpressionBase> list = _bindings.ToList();
            for/*each*/ (BindingExpressionBase binding in list) 
            {
                // fetching TargetElement may detach the binding, removing it from _bindings 
                DependencyObject owner = binding.TargetElement; 
            }
        } 
        foundDirt = foundDirt || (_bindings.Count < count);

        count = _bindingGroups.Count;
        if (count > 0) 
        {
            List<BindingGroup> list = _bindingGroups.ToList(); 
            for/*each*/ (BindingGroup bindingGroup : list) 
            {
                // fetching Owner may detach the binding group, removing it from _bindingGroups 
                DependencyObject owner = bindingGroup.Owner;
            }
        }
        foundDirt = foundDirt || (_bindingGroups.Count < count); 

        return foundDirt; 
    } 

//    #endregion Internal Methods 

//    #region Private Methods

    // return true if element is a descendant of ancestor 
    boolean IsInScope(DependencyObject ancestor, DependencyObject element)
    { 
        boolean result = (ancestor == null) || VisualTreeHelper.IsAncestorOf(ancestor, element); 
        return result;
    } 

//    #endregion Private Methods

//    #region Private Data 

    Set<BindingGroup> _bindingGroups = new Set<BindingGroup>(); 
    Set<BindingExpressionBase> _bindings = new Set<BindingExpressionBase>(); 

    static final List<BindingGroup> EmptyBindingGroupList = new List<BindingGroup>(); 
    static final List<BindingExpressionBase> EmptyBindingList = new List<BindingExpressionBase>();

//    #endregion Private Data

//    #region Private types

    class Set<T> extends Dictionary<T, Object> implements IEnumerable<T> 
    {
        public Set() 
            : base()
        {
        }

        public Set(IDictionary<T, Object> other)
            : base(other) 
        { 
        }

        public Set(IEqualityComparer<T> comparer)
            : base(comparer)
        {
        } 

        public void Add(T item) 
        { 
            if (!ContainsKey(item))
            { 
                Add(item, null);
            }
        }

        IEnumerator<T> IEnumerable<T>.GetEnumerator()
        { 
            return Keys.GetEnumerator(); 
        }

        public List<T> ToList()
        {
            return new List<T>(Keys);
        } 
    }

//    #endregion Private types 
}